### PR TITLE
Setup Github Actions

### PR DESCRIPTION
**Pull request by jameesjohn**
## Explanation

Setup github actions.
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by jameesjohn** [comment]
PTAL @ankita240796.
Thanks!

**Comment by jameesjohn** [comment]
> Is the file dist/index.js generated after build? 

Yes.

> Also, what is the purpose of the file? Can we gitignore it and build everytime it is required?

Since it is an action to be deployed, we can either commit the node_modules directory which will contain the github/core and github/github packages that are needed for actions to run or we build the file beforehand and commit the built file.
One of these two need to be done.

**Comment by jameesjohn** [comment]
This feature was tested at https://github.com/jameesjohn/comment-on-pr/issues/10

**Comment by ankita240796** [comment]
> > Is the file dist/index.js generated after build?
> 
> Yes.
> 
> > Also, what is the purpose of the file? Can we gitignore it and build everytime it is required?
> 
> Since it is an action to be deployed, we can either commit the node_modules directory which will contain the github/core and github/github packages that are needed for actions to run or we build the file beforehand and commit the built file.
> One of these two need to be done.

Is this the behaviour followed by other repos too? Can we look into some examples here?

Also, let us split this PR. Have only the setup stuff here and move the checks into a PR stacked over this one.

**Comment by jameesjohn** [comment]
> Is this the behaviour followed by other repos too

Yes. This is the approach recommended by GitHub. https://help.github.com/en/actions/building-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github.

**Comment by ankita240796** [comment]
> > Is this the behaviour followed by other repos too
> 
> Yes. This is the approach recommended by GitHub. https://help.github.com/en/actions/building-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github.

Sg! Let us split the PR in that case. Thanks!

**Comment by jameesjohn** [comment]
> One other doubt: what happens when we update packages - we will need to rebuild and push?

Yes.

**Comment by jameesjohn** [comment]
> Sg! Let us split the PR in that case. Thanks!

Done, PTAL. Thanks!

**Comment by jameesjohn** [comment]
>  can we have an automated test to ensure re-build is done when any package is updated?

Done.

> What is the purpose of the whitelist.json file?

It is to hold users that are to be whitelisted from different checks.

> What are the next steps after this PR?

To setup issues labeled checks then we publish

> Also, do we need to setup all the checks in the actions folder to be able to publish them or only the setup part as done in this PR?

That is what I intend to do to prevent mixing actions specific code from probot code.

File: .circleci/config.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ workflows:
   circleci_tests:
     # The following jobs run in the following stages:
     # - Setup runs first.
-    # - After setup passes, frontend tests and coverage tests run.
+    # - After setup passes, frontend tests, coverage tests and actions tests run.
```
Comment (ankita240796) [inline .circleci/config.yml:7]: Revert.

Comment (jameesjohn) [reply .circleci/config.yml:7]: Done.

File: actions/build.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+const { execSync } = require('child_process');
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply actions/build.js:1]: Done.

File: actions/src/dispatcher.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  LABELLED: 'labeled'
+};
+module.exports = {
+  async dispatch(event, action) {
+    core.info(`Received Event:${event} Action:${action}.`);
+    if (event === EVENTS.ISSUES) {
+      if (action === ACTIONS.LABELLED) {
+        // Handle Labelled Action
+      }
+    }
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply actions/src/dispatcher.js:34]: Removed.

File: package.json [hunk 4]
```diff
@@ -46,5 +54,15 @@
     "watch": [
       ".env"
     ]
+  },
+  "lint-staged": {
+    "actions/**/*.+(js|json)": [
+      "npm run update-actions-build"
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply package.json:60]: We are using [husky](https://www.npmjs.com/package/husky) for the pre-commit check. On line 65, we tell husky to call lint-staged before a commit is made.
Then [lint-staged](https://www.npmjs.com/package/lint-staged) checks if the changed files pass the regex defined in line 59 (if a file in the actions directory was changed) and calls `npm run update-actions-build` which rebuilds the actions file and automatically adds it to the commit.

File: whitelist.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{}
```
[the remark this answers is not in the document]
Comment (jameesjohn) [reply whitelist.json:1]: True.
I'll be removing it from here.

**Comment by jameesjohn** [comment]
PTAL @ankita240796, I've addressed all review comments.

File: actions/build.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+// limitations under the License.
+
+/**
+ * @fileoverview Automatically compile github actions when
+ * actions related file changes.
+ */
+
+const { execSync } = require('child_process');
+
+// Build the actions file
```
Comment (ankita240796) [inline actions/build.js:22]: Period in end.

File: actions/build.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+execSync('npm run actions-build');
+// Automatically add the build file to the commit index
```
Comment (ankita240796) [inline actions/build.js:24]: Ditto.

File: spec/actionsBuild.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
Comment (ankita240796) [inline spec/actionsBuild.js:1]: What about this? Did you try the regex based check that we discussed?

Comment (jameesjohn) [reply spec/actionsBuild.js:1]: I'm getting a weird error from while getting the diff.
Here's the error log: https://pastebin.com/yb6iqnUY

The command I'm running is `git diff --no-index dist_file_path new_build_file` from node.
Running this normally gives the expected output

File: actions/build.js (hunk 6 above)
Comment (jameesjohn) [reply actions/build.js]: Done.

File: actions/build.js (hunk 7 above)
Comment (jameesjohn) [reply actions/build.js]: Done.

**Comment by jameesjohn** [comment]
Hi @ankita240796, before we get this in, we need to choose a feather icon to use as the logo for the action from https://feathericons.com/ and also we need to choose the colour of the logo. The accepted colours are white, yellow, blue, green, orange, red, purple, or gray-dark.

File: package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ankita240796) [reply package.json:60]: Thanks for the explanation.

File: spec/actionsBuild.js (hunk 8 above)
Comment (ankita240796) [reply spec/actionsBuild.js:1]: Hmm, did you look into what can be causing the error? Can we try adding a catch block to the promise to handle any errors like this? One other way could be to use a custom function which accepts two string args and finds the diff between them.

**Comment by jameesjohn** [comment]
> Regarding your query for icons, is it possible to use custom icons/images i.e. the one we already use for oppiabot?

No that is not supported for GitHub actions https://github.community/t/can-one-set-a-custom-image-icon-for-a-github-action-or-github-workflow-or-only-github-apps/17513/2

**Comment by ankita240796** [comment]
> > Regarding your query for icons, is it possible to use custom icons/images i.e. the one we already use for oppiabot?
> 
> No that is not supported for GitHub actions https://github.community/t/can-one-set-a-custom-image-icon-for-a-github-action-or-github-workflow-or-only-github-apps/17513/2

Hmm ok, in that case - I do not have any strong preference. We can go with something common which is being used by other projects as well.

File: spec/actionsBuild.js (hunk 8 above)
Comment (jameesjohn) [reply spec/actionsBuild.js]: Following discussion via chat, we resolved to stick to just the pre commit hook.

**Comment by jameesjohn** [comment]
@ankita240796, I used the purple as the color and activity as the icon.

Here's what it will look like:
Screenshot (413)

**Comment by ankita240796** [comment]
> @ankita240796, I used the purple as the color and activity as the icon.
> 
> Here's what it will look like:
> Screenshot (413)

This looks good!

**Pull request merged**